x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/pivot/DateHistogramGroupSource.java

```diff
@@ -88,6 +97,70 @@ public Type getType() {
         return Type.DATE_HISTOGRAM;
     }
 
+    @Override
+    public QueryBuilder getNextBucketsQuery(Object key) {
+        long bucketKey = parseBucketKey(key);
+        return QueryBuilders.rangeQuery(field).gte(getNextBucketKey(bucketKey));
+    }
+
+    @Override
+    public QueryBuilder getCurrentBucketQuery(Object key) {
+        long bucketKey = parseBucketKey(key);
+        return QueryBuilders.rangeQuery(field).lt(getNextBucketKey(bucketKey)).gte(bucketKey);
+    }
+
+    private long parseBucketKey(Object key) {
+        // The user can optionally set the format to make the typically millisecond number more human readable
+        // So, if the user set the format, parse the bucket key according to that format
+        if (this.format != null) {
+            DateFormatter formatter = DateFormatter.forPattern(this.format);
+            return DateFormatters.from(formatter.parse((String)key)).toInstant().toEpochMilli();
+        } else {
+            assert key instanceof Number;
+            return (Long)key;
+        }
+    }
+
+    private long getNextBucketKey(long interval) {
+        return getRounding().nextRoundingValue(interval);
+    }
+
+    private Rounding getRounding() {
```

Comment:
These rounding methods are basically the same as used in the DateHistogramAgg class.